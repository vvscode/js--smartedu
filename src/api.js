import axios from "axios";

const API_BASE = "https://django-smartedu.herokuapp.com/api";

export const getCourses = () =>
  axios.get(`${API_BASE}/courses/`).then((response) => response.data);

export const getCourse = (id) =>
  axios.get(`${API_BASE}/courses/${id}`).then((response) => response.data);

export const getGroup = (id) =>
  axios.get(`${API_BASE}/group/${id}`).then((response) => response.data);

export const sendApplication = (applicationDetails) =>
  axios
    .post(`${API_BASE}/application`, applicationDetails)
    .then((response) => response.data);

export const getMyInfo = () =>
  axios.get(`${API_BASE}/users/me/info`).then((response) => response.data);

export const getUserCourses = () =>
  axios.get(`${API_BASE}/users/me/courses/`).then((response) => response.data);

export const regenerateToken = (currentToken) =>
  axios
    .post(`${API_BASE}/users/me/regenerate-token`, currentToken)
    .then((response) => response.data);
