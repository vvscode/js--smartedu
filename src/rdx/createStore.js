import { configureStore } from "@reduxjs/toolkit";
import userSlice from "./user";

const reducer = {
  user: userSlice.reducer,
};

export const createStore = (preloadedState) =>
  configureStore({
    reducer,
    preloadedState,
  });
