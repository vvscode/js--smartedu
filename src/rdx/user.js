import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { getMyInfo } from "../api";

export const fetchCurrentUserInfo = createAsyncThunk(
  "user/fetchCurrentUserInfo",
  async () => {
    return await getMyInfo();
  }
);

export const selectUserInfo = (state) => state.user.userInfo;

export default createSlice({
  name: "user",
  initialState: {
    userInfo: null,
  },
  reducers: {
    [fetchCurrentUserInfo.fulfilled.type]: (state, action) =>
      (state.userInfo = action.payload),
    [fetchCurrentUserInfo.rejected.type]: (state) => (state.userInfo = null),
    deleteUserInfo: (state) => (state.userInfo = null),
    updateUserToken: (state, action) =>
      (state.userInfo = { ...state.userInfo, token: action.payload }),
  },
});
