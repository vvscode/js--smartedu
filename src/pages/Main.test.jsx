import React from "react";
import MainPage from "./Main";
import * as api from "../api";
import { render } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router";
import { createStore } from "../rdx/createStore";
import { Provider } from "react-redux";

jest.mock("../api");

const makeCourse = (id) => ({
  id,
  title: `Title for course #${id}`,
  description: `Description for course #${id}`,
  tags: [...new Array({ length: id })].map(
    (el, index) => `Tag ${index} for course #id`
  ),
});

describe("Main Page", () => {
  let store;
  const renderMainPage = () => {
    store = createStore();
    render(
      <Provider store={store}>
        <Router>
          <MainPage />
        </Router>
      </Provider>
    );
  };

  it("requests new courses on render", () => {
    api.getCourses.mockResolvedValueOnce([]);
    renderMainPage();
    expect(api.getCourses).toHaveBeenCalled();
  });

  it("renders loaded courses", async () => {
    api.getCourses.mockResolvedValueOnce([makeCourse(1), makeCourse(2)]);
    const wrapper = renderMainPage();
    await api.getCourses();
    expect(wrapper).toMatchSnapshot();
  });
});
