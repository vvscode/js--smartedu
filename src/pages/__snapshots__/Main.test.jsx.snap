// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page renders loaded courses 1`] = `undefined`;
